File: evaluateReversePolishNotation.py
# problem: https://leetcode.com/problems/evaluate-reverse-polish-notation/
# Runtime: 60 ms, faster than 92.42% of Python3 online submissions for Evaluate Reverse Polish Notation.
# Memory Usage: 14.7 MB, less than 42.12% of Python3 online submissions for Evaluate Reverse Polish Notation.

from typing import List
from math import ceil


class Solution:
    def evalRPN(self, tokens: List[str]) -> int:

        stack = []
        for token in tokens:
            if token.lstrip("-").isnumeric():
                stack.append(int(token))
            else:
                a = stack.pop()
                b = stack.pop()
                if token == "*":
                    res = b * a
                elif token == "/":
                    res = b // a if b / a > 0 else ceil(b / a)
                elif token == "+":
                    res = b + a
                else:
                    res = b - a
                stack.append(res)
        return stack[0]

File: evaluateReversePolishNotation.cpp
// problem:
// https://leetcode.com/problems/evaluate-reverse-polish-notation/description/
// Runtime
// 11 ms
// Beats
// 89.4%
// Memory
// 11.9 MB
// Beats
// 93.66%

#include <iostream>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
  int evalRPN(vector<string> &tokens) {
    vector<long> nums;
    for (auto const &token : tokens) {
      if (token == "+") {
        long first = nums[nums.size() - 1];
        nums.pop_back();
        long second = nums[nums.size() - 1];
        nums.pop_back();
        nums.push_back(first + second);
      } else if (token == "-") {
        long first = nums[nums.size() - 1];
        nums.pop_back();
        long second = nums[nums.size() - 1];
        nums.pop_back();
        nums.push_back(second - first);
      } else if (token == "*") {
        long first = nums[nums.size() - 1];
        nums.pop_back();
        long second = nums[nums.size() - 1];
        nums.pop_back();
        nums.push_back(first * second);
      } else if (token == "/") {
        long first = nums[nums.size() - 1];
        nums.pop_back();
        long second = nums[nums.size() - 1];
        nums.pop_back();
        nums.push_back(second / first);
      } else {
        nums.push_back(stoi(token));
      }
    }
    return nums[0];
  }
};
